### ISSUE TITLE
Running sqlboiler requires basedir for templates to be present on disk

### ISSUE THREAD

**Issue by spenczar** [volatiletech/sqlboiler#151]
Hello! I'm coming to sqlboiler not as a user, but as someone who builds other tools. sqlboiler was mentioned in https://github.com/twitchtv/retool/pull/25.

sqlboiler looks seriously great, but the fact that it requires the existence of the `templates` and `templates_test` directories (and their contents) limits the portability of its binary, and means other tools have to do funny contortions.

https://github.com/jteeuwen/go-bindata is a tool that could be used to alleviate this. Through code generation, it will bake all the templates directly into the binary (in compressed, gzipped form) and provide accessors for them. Instead of reading files, you'd fetch them with calls like `raw, err := Asset("templates/some_template.tpl")`. Similarly, you can iterate over directories, with `AssetDir("templates")`.

Just a suggestion - this could make sqlboiler much more portable and friendly to use with other tools.

**Comment by aarondl** [comment]
Hey there. We're already aware of go-bindata and had previously made the choice not to use it mostly because it just makes more work for us and no one's complained loud enough yet.

We talked about it briefly and felt like we can definitely support this workflow. We're talking about doing this in two different ways:
- Mandatory and always-on, this has the downside that rapid iteration on templates now requires go-bindata & go builds in between which frankly is not awesome.
- Optional and controlled by build tags (leaning towards this one). This allows `go get` to continue to work which I think most people are using still.

Currently there's a way to do this by bundling the template files with your app and supplying `--basedir`. SQLBoiler doesn't need it's source so much as there's a convenience mechanism that `--basedir` is autoset to it's place in `$GOPATH` to find the official templates, so you could work around this like this for now though I'm not sure if that truly fits the use case.

We'll probably implement this, but in the interest of setting expectations I want to say that it's not high on the feature list that we want to get done.

Hope that helps.

**Comment by Alexendoo** [comment]
For the rapid iteration point go-bindata has a debug build mode that doesn't require generation or rebuilds for file changes

**Comment by spenczar** [comment]
Yep, I understand - it's definitely a balance between your convenience versus others when working with this project.

@Alexendoo mentioned a good middle ground, which is the debug mode for go-bindata, although you have to remember to re-generate without debug turned on before committing to master. Another option is to keep the `--basedir` option (probably renaming it) as an override to use templates from a different directory. You would want this anyway for backwards compatibility. You could use it when making sure your changes work.

You could set this when running tests, as well, if that's part of your workflow when working on templates.

Are you suggesting build tags as a way to opt-in to portability, or a way to opt-out? If it's opt-in, then it doesn't gain a _whole_ lot from my perspective. Tools like retool would still need to know to specially set that build tag. But opt-out makes sense.

**Comment by aarondl** [comment]
I guess I'm not understanding what you guys are after then. To me having this be opt-in means that we can start publishing binaries alongside the Github releases. We originally did this then realized it was stupid without bindata and stopped. Retool just downloads this binary and uses it.
And then of course making things opt-in keeps backwards compat, ensures we commit no generated artifacts to the repository which is also great from a "cleanliness" perspective.

So what I'm proposing is that for use cases where a single binary is desirable a released version is available for all consumers (including retool) which will be built by our CI and published to Github using
an an opt-in build process that shields existing/regular users from the awkwardness of go-bindata using build-tags. This means it's impossible using `go get` workflows to "forget" or
have a mismatch of templates. It keeps a generated duplication of the templates in another form out of the repository.

If this is acceptable for you then I'd say we go this way. If this completely doesn't work for you then we can still consider the Mandatory alway-on as another option but I don't think opt-out is desirable as I think it throws away most of the advantages of having it be toggle-able.

**Comment by spenczar** [comment]
Ah, I think we are talking about slightly different things, but are interested in the same goal, really: we both want sqlboiler to work well with `go get`.

`retool` in fact just uses `go get`, but in a temporary `GOPATH`. It expects that it can build the binary with a plain, vanilla `go get`, and then use that binary without keeping around the source files. 

To me, using go-bindata (or whatever equivalent) is actually a way to make the thing _more_ `go get`-friendly, since it doesn't require maintaining the same state of the GOPATH after you've run `go get`. It's also more predictable, since it doesn't change behavior until you rebuild it.

Thanks for working with me on this, by the way! I'm pretty sure we're making progress towards something that works on both sides.

**Comment by spenczar** [comment]
Two binaries is not a problem. Just as you ask your users to run two `go get` commands, retool users would run two `retool add` commands. I think that would be fairly clear to retool's users.

I don't see why retool would need to consume binaries directly in order for it to work with the driver split, could you elaborate?

**Comment by Alexendoo** [comment]
In regards to simplicity I would ask how much size you're saving by having a shared binary dependency on sqlboiler, currently it's 13M for the one including all the core drivers, most people would only use 1 or 2 drivers at most so perhaps it would be worth driver binaries depending on sqlboiler as a library rather than calling it as a binary at runtime.

If you split into a different repo called sqlboiler-lib or something it would be a good way to provide backwards compatibility also for existing sqlboiler users.

**Comment by aarondl** [comment]
Ah I misunderstood. I didn't know that `retool` just keeps binaries on disk. I thought it was more like a use case we heard before where the binaries are bindata'd inside `retool`. You're right, you'd just have to `go get` N times - one for each driver. Though I still think that `retool` should support fetching binaries ;)

@Alexendoo Both "side effect import" style inclusion as well as the "separate binary" style is planned. We use sqlboiler as a library in another project so we want to keep that functionality. If we want to discuss this issue let's take it to https://github.com/vattle/sqlboiler/issues/143

**Comment by aarondl** [comment]
Thinking about this off and on all day. We have a few options:

### SQLBoiler

1. Mandatory go-bindata for SQLBoiler

    - Only SQLBoiler changes
    - `go get` continues to work
    - Dev workflow is strictly worse
    - Allows released binaries to exist with sqlboiler again

1. Opt-in go-bindata

    - Allows current workflows to remain un-changed
    - Allows released binaries to exist with sqlboiler again
    - Downside is it requires build flags (+bindata)

### Retool

1. Consume binaries from github releases or anything else

    - Example: `retool add-binary version URL`
    - Allows retool to bundle ANY binary, not just Go binaries (which is actually awesome)

1. Run `go generate` as a step when getting new go code

    - Allows retool to build any required generated pieces when downloading source code
    - Allows sqlboiler to use `go generate` as a contract to ensure all pieces are ready for compilation (avoids makefiles and other complicated pieces)
    - Would require retool to also always supply the `+bindata` flag as an interface to `go get` (or something else)

My preference on middleground would be SQLBoiler 2, and Retool 2. Though I think Retool 1 has some serious awesomeness (as a potential user of retool - I like the idea of it and I like being able to treat binaries as binaries not Go source code even more).

**Comment by spenczar** [comment]
Thanks for the concrete plans, @aarondl!

Regarding the retool things:

We don't consume binaries directly because it's not possible to do so in a good cross-platform way, and it's not easy to do with dynamic linking in binaries, so the tool becomes very unpredictable in its behavior. The core benefit of retool is reproducibility, so this is crucial. It's also not clear how this would work in a no-network environment, like in build boxes.

Retool shouldn't be running `go generate` itself because it's too much of a security risk. Doing so would mean you run arbitrary code from a remote source.

Regarding the sqlboiler things:

I think you are overestimating the difficulties in using `go-bindata` as a developer. If it feels bad to run `go generate` after making edits to templates, you can always write a Makefile or a script to run `go generate` before running `go test`. You can go further by doing things like checking an environment variable during tests, and if it's present, then don't read the generated bindata, and instead use the template files on disk. Then, the dev workflow just entails exporting that environment variable before doing work. There are plenty more ways to make things smooth.

**Comment by aarondl** [comment]
Consuming binaries in a cross platform way should be very simple right? After all you're going to be downloading source from some public URL anyway, you simply divide them into the different `GOOS`/`GOARCH` in some yaml lock file like you've got now, and then download the appropriate one since you can query for those variables inside a Go program (like `retool`). Also consider the many other repos that are not go-gettable and require make files and what-not, how does retool deal with these today? Wouldn't this broaden the amount of tools `retool` could work with by a lot?

It's also fairly clear how it would work inside a no-network environment like a build box since the source code has to be available and was previously fetched, the binaries could all be fetched instead. Though this does of course bloat the payload, but unless you're passing it around on a 128MB USB stick it should be okay.

I'd also mention that it's -more- reproducible to pull a binary, since you're building binaries with a potentially different Go version than it would have originally been built with (maybe something doesn't work with Go 1.7+ for example, but you're compiling with Go 1.8).

From what you've described/what I've seen of retool, dynamic linking is already a problem?
- How do you know what libs (and which versions) to download for a given source code base, and how do you get them in a cross platform way (`apt-get`, `zypper`, `yum`, `pkgsrc`, `apk`, `pacman` and of course the lib name which is never the same).
- Compiling cgo requires gcc/mingw installed, it's possible that that's not available on a system where Go is installed running `retool`

Unless we're talking like libc resolver problems that Go has inside of alpine because musl or something. Though this is now solvable at runtime afaik (https://golang.org/pkg/net/#hdr-Name_Resolution).

Given the above I don't think that consuming binaries is as bad as you think, and the only problem lies in potential security updates to Go itself, but then again `retool` isn't really meant for running http servers on production machines as much as packaging development tools, and that wouldn't suffer from this, correct?
Though I agree that `go generate` is a bad way to go because of the execution of arbitrary code on build systems so we can rule that one out.

There's no real problem with adding `bindata` from a technical perspective. Sure it's easy to manually run `go generate` or try to remember to use a `Makefile` that does the extra steps. But that's the problem I have with the change philosophically. We're introducing manual steps that aren't taken care of by CI, and I simply don't appreciate any workflow that's like that. There's a reason we build & run the tests automatically after a push, because we may have forgotten. We're squishy bags of mistake-meat. Just like we may have forgotten to re-generate the bindata and are running against old code when we run our tests. Forgetting to set the dev environment env var. It's just "more things". When you arrive in a Go project `go get`, `go build`, `go test`, these are your typical tools, if you have to do more than that that means more documentation, and a higher possibility of screwing it up. Also I've worked in many projects with `go-bindata` and it's never as straightforward as you make it out to be, it definitely adds to the maintenance burden, it seems to always find a way to bite you at some moment you don't expect it.

Now you're right in that I'm overblowing it a tad though, it's the slippery slope fallacy. But at least forgive me for trying to seek out EVERY possible alternative to this - seemingly inevitable - solution of just straight-up using bindata. I was *really* hoping it wouldn't come to this and we could do something nice for everyone involved.

**Comment by Alexendoo** [comment]
For the retool stuff: my use case is fairly straightforward, I want to be able to have anybody working on a project be able to use the same codegen/test tools. I've yet to encounter a tool in the go environment that isn't simply `go get`able, so there's no problem for me there

Retool makes this super easy, I can just `retool add ...` and commit the config file, then anybody using retool also can have the tools installed in a single command.

Vendoring direct binaries would be sufficiently more difficult on the maintainer, they would have to manually add links to the tools.json for each OS/Arch and manually change them for every upgrade, and then recalculate the hashes (which could be done semi automatically)

-----

For the forgetting to regenerate thing, yeah I've been there a couple times. You can set up CI to catch that for you - install the needed tools (with retool :D) and do a go generate, if git sees any files changed fail the build

**Comment by aarondl** [comment]
I understand the use case, trust me! Since this issue came up I actually want to use `retool` since I've seen the issues it solves, I think it's great :D Also I've been around the block a few times and have seen tools that require make files or simply suck at building correctly (I'm looking at you go-swagger). It's reasonably common to have a make file or other build process inject an `ldflags` argument to the compiler to embed version information that's not hardcoded into the source code as an example.

I'm not sure why they would have to manually add any links though to `tools.yml`, it should be simple to make a `retool add-binary` as I mentioned before that could do the exact same thing as grabbing the sources. It could even try to be smart and find the other arches by munging the URL (pretty standard to see linux-amd64 or some such combination near the end), and/or classifying it automatically and prompting the user if it couldn't figure out what os/arch the binary is from the name. This is of course assuming that it's important to get each os/arch, which it's probably not.

For upgrading we could potentially use the github API to figure out the latest release on some branch (or the latest release in general). https://developer.github.com/v3/repos/releases/

In the end of all of this however, I think I'm ready to capitulate. I simply think the energy going into this isn't worth it anymore. I think I've presented some pretty reasonable arguments for why both `retool` and `sqlboiler` should be changed in the ways I described above to the benefit of everyone, but I don't seem to be gaining ground much to my dismay.

So sqlboiler will adopt go-bindata. However, I think this is very far out from completion as I need to release both authboss v2 and sqlboiler v3 in the next little while and am working on the features for those (and even those are going molasses-slow lol). I understand from the PR that was linked above that there's a workaround currently for retool so it's not urgent anyway, is this correct?
